compressors/arithmetic_coding.py

```diff
@@ -0,0 +1,436 @@
+        return s
+
+    def decode_block(self, encoded_bitarray: BitArray):
+
+        data_block_size_bitarray = encoded_bitarray[: self.params.DATA_BLOCK_SIZE_BITS]
+        encoded_bitarray = encoded_bitarray[self.params.DATA_BLOCK_SIZE_BITS :]
+
+        # get data size
+        input_data_block_size = bitarray_to_uint(data_block_size_bitarray)
+
+        # initialize return variables
+        decoded_data_list = []
+        num_bits_consumed = 0
+
+        # initialize intermediate state vars etc.
+        low = 0
+        high = self.params.FULL
+        state = 0
+        arith_bitarray_size = len(encoded_bitarray)
+
+        # initialize the state
+        while (num_bits_consumed < self.params.PRECISION) and (
+            num_bits_consumed < arith_bitarray_size
+        ):
+            bit = encoded_bitarray[num_bits_consumed]
+            if bit:
+                state += 1 << (self.params.PRECISION - num_bits_consumed - 1)
+            num_bits_consumed += 1
+        num_bits_consumed = self.params.PRECISION
+
+        # main decoding loop
+        while True:
+            # decode the next symbol
+            s = self.decode_step_core(low, high, state, self.freq_model.freqs_current)
+            low, high = ArithmeticEncoder.shrink_range(self.freq_model.freqs_current, s, low, high)
+            decoded_data_list.append(s)
+
+            # update the freq_model
+            self.freq_model.update_model(s)
+
+            # break when we have decoded all the symbols in the data block
+            if len(decoded_data_list) == input_data_block_size:
+                break
+
+            while (high < self.params.HALF) or (low > self.params.HALF):
+                if high < self.params.HALF:
+                    # re-adjust range, and reset params
+                    low = low << 1
+                    high = high << 1
+                    state = state << 1
+
+                elif low > self.params.HALF:
+                    # re-adjust range, and reset params
+                    low = (low - self.params.HALF) << 1
+                    high = (high - self.params.HALF) << 1
+                    state = (state - self.params.HALF) << 1
+
+                if num_bits_consumed < arith_bitarray_size:
+                    bit = encoded_bitarray[num_bits_consumed]
+                    state += bit
+                num_bits_consumed += 1
+
+            while (low > self.params.QTR) and (high < 3 * self.params.QTR):
+                # increment the mid-range adjustment counter
+                low = (low - self.params.QTR) << 1
+                high = (high - self.params.QTR) << 1
+                state = (state - self.params.QTR) << 1
+
+                if num_bits_consumed < arith_bitarray_size:
+                    bit = encoded_bitarray[num_bits_consumed]
+                    state += bit
+                num_bits_consumed += 1
+
+        # # NOTE: we might have loaded in additional bits not added by the arithmetic encoder
```

Comment:
@kedartatwawadi - I think the understand the whole thing except for this last part. If you can explain it here or in the wiki, I'm happy to approve this.